### ISSUE TITLE
[Bug Report] Implicit row broadcast causes accuracy drop in backward pass for Llama LoRA 1B

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-metal#27676]
### Describe the bug

```use_legacy = False``` is not default yet for binary ops where row broadcast is needed for one of the operands and so if such op is invoked through Forge the accuracy will suffer.
Manually enabling ```use_legacy = False``` in tt-mlir runtime resolved this accuracy issue. 

Repro is given below for this specific Llama LoRA 1B use case, where the accuracy problem manifests as pushing smaller values to 0.

### Steps to reproduce the issue

Run:
```python
import ttnn
import torch

with ttnn.manage_device(device_id=0) as device:

    # Scaling with 0.01 to get realistic values that appear during training.
    a = ttnn.from_torch(0.01 * torch.randn((1, 1, 2048), dtype=torch.float32), layout = ttnn.TILE_LAYOUT, device=device)
    b = ttnn.from_torch(0.01 * torch.randn((4, 63, 2048), dtype=torch.float32), layout = ttnn.TILE_LAYOUT, device=device)
    
    out_torch = torch.multiply(ttnn.to_torch(a), ttnn.to_torch(b))
    print("Torch multiply result:")
    print(out_torch)

    out = ttnn.multiply(a, b)
    print("TTNN multiply result:")
    print(out)
```
to get:
```
Torch multiply result:
tensor([[[ 7.8362e-05,  4.6404e-06,  1.0304e-04,  ...,  2.4088e-05,
          -6.1180e-06, -3.0956e-06],
         [-2.1834e-05, -3.4601e-06, -3.8998e-05,  ..., -1.5977e-05,
           3.8527e-06, -5.2687e-07],
         [ 4.5182e-05, -8.0412e-06, -7.7680e-05,  ..., -4.3809e-07,
           1.5390e-06,  2.9570e-06],
         ...,
         [-7.6850e-05, -2.6978e-05,  2.1537e-05,  ...,  3.0352e-05,
          -8.6119e-06, -1.6524e-06],
         [-1.5798e-04, -1.0613e-05, -3.7203e-05,  ...,  5.0910e-05,
           5.4463e-06,  1.5863e-06],
         [ 3.5050e-05,  1.9119e-05, -1.7322e-05,  ..., -1.8598e-05,
           1.7699e-05,  2.8308e-06]],

        [[ 2.2873e-05,  3.8722e-07, -5.7004e-05,  ...,  3.8156e-05,
          -2.4726e-06,  1.3411e-06],
         [ 5.6266e-06,  2.7303e-05, -8.6021e-05,  ...,  3.3704e-05,
          -1.9897e-06, -2.6866e-06],
         [ 4.1571e-05,  8.4202e-06,  7.6078e-05,  ..., -1.3991e-06,
          -1.7463e-05, -2.4630e-06],
         ...,
         [ 1.6496e-05, -2.2934e-05, -7.4409e-06,  ...,  3.8968e-06,
           2.1250e-05, -1.0590e-06],
         [ 7.1580e-05, -1.2252e-05,  4.5608e-05,  ..., -2.4936e-05,
           8.6773e-06,  2.8771e-06],
         [ 8.2760e-05, -1.0785e-05,  1.8031e-05,  ...,  2.4266e-05,
           1.9352e-05,  6.2956e-07]],

        [[ 1.6364e-05,  1.5900e-05, -8.5472e-05,  ..., -2.1448e-08,
           7.7940e-06, -1.7159e-06],
         [ 7.5317e-05,  1.5246e-05,  6.5830e-05,  ..., -2.8578e-06,
           1.0964e-05,  1.9375e-07],
         [ 1.6771e-04, -2.7158e-06, -2.9145e-05,  ..., -5.1924e-05,
          -2.6828e-07,  8.9363e-07],
         ...,
         [ 5.1816e-05, -1.5456e-05, -8.2151e-05,  ...,  4.3050e-05,
          -4.4231e-06, -2.0345e-07],
         [-2.7687e-05,  1.8909e-05, -6.7756e-05,  ...,  6.3069e-05,
           8.7283e-06, -9.6685e-07],
         [-1.4172e-05, -5.7652e-06, -3.7233e-05,  ..., -8.5172e-05,
          -5.5707e-07, -1.2951e-06]],

        [[ 6.0911e-05,  1.9704e-06, -1.3554e-05,  ..., -3.2856e-05,
           1.0093e-05,  7.5925e-07],
         [ 1.8818e-05,  2.2879e-05,  6.0145e-05,  ..., -2.4351e-05,
           1.0312e-05, -5.1344e-07],
         [ 4.8041e-05, -1.8760e-05, -6.1016e-05,  ..., -1.7358e-06,
          -3.2998e-07,  2.5188e-06],
         ...,
         [-1.7644e-05, -3.1527e-05,  3.1491e-05,  ...,  2.5113e-06,
           2.1487e-05,  1.6026e-07],
         [-3.2172e-05, -8.9557e-06,  1.7777e-05,  ...,  3.6902e-05,
          -1.4277e-05, -5.8635e-07],
         [ 6.4867e-05,  4.4120e-06, -8.9535e-05,  ..., -6.2143e-06,
          -5.5051e-06,  2.5221e-07]]])
TTNN multiply result:
ttnn.Tensor([[[ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              ...,
              [-0.00015,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000]],

             [[ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              ...,
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000]],

             [[ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              ...,
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000]],

             [[ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              ...,
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000],
              [ 0.00000,  0.00000,  ...,  0.00000,  0.00000]]], shape=Shape([4, 63, 2048]), dtype=DataType::FLOAT32, layout=Layout::TILE)
```

### Expected behavior

```use_legacy = False``` should be made default for implicit row broadcast case so that ops that lower through Forge onto such cases are executed without accuracy issues.

### Please complete the following environment information

- Version of software (commit hash) https://github.com/tenstorrent/tt-metal/commit/f34fc81cf24421c220b0d8255f21b036968390bf

**Comment by cmaryanTT** [comment]
@dchenTT - note that this is fp32

**Comment by dchenTT** [comment]
This is the trade off. For fp32, binary_ng uses software subtile broadcast therefore it is slow. For legacy binary, it uses a version of LLK therefore it is fast but it uses FPU and internally it is TF32. Currently performance takes precedence, so row broadcast is directed to legacy. The exception is for bf16, it is by binary_ng that is perfect. Currently we have dependency on LLK team so we are blocked to let binary_ng handle fp32 with fast performance.

For now, there are two choices for Forge folks. Fast performance but it is actually TF32 internally. Or slow performance with full fp32 accuracy. We cannot have both in the same time. The best way is still for Forge folks to add the parameter of use_legacy, so that it has the freedom to choose.

**Comment by dchenTT** [comment]
Unless we make adjustment that for fp32, accuracy takes precedence so it will be default to binary_ng. However, I am not sure existing performance related CI will break or not. If not, we may do so.

**Comment by cmaryanTT** [comment]
My general standard is that any data movement, like broadcast, should preserve full accuracy by default. So I'd like to have the default accuracy be fp32 in this case.

That said, I don't think this is a TF32 problem. The input tensors have values with a mean of 0.5e-3, so mean of the output should be around 0.25e-6 as in the torch example. That's comfortably within the tf32 range (1e-38 I believe). So something else is going on.

**Comment by dchenTT** [comment]
Then it should have something to do with FPU vs SFPU. Legacy binary does not use SFPU with broadcasting. Instead of fix or improve legacy, I will create a PR to direct to binary_ng for fp32 for row broadcast.

**Comment by dchenTT** [comment]
https://github.com/tenstorrent/tt-metal/pull/27794 merged.

**Comment by cmaryanTT** [comment]
@umalesTT - can you please validate on your end, and then we can close this

**Comment by umalesTT** [comment]
@cmaryanTT this metal change still isn't uplifted in tt-xla (the repo we are running code from), so we can't really validate it yet. I'll validate it once the change is uplifted. If we don't want to keep the issue open in the meantime, I think we can close this because this particular repro is really 1-1 with our frontend and if it works here there is no reason why it wouldn't work on frontend.
